src/content/learn/typescript.md

```diff
@@ -0,0 +1,433 @@
+
+function MyComponent() {
+  const object = useGetComplexObject();
+
+  return (
+    <div>
+      <p>Current object: {object.kind}</p>
+    </div>
+  )
+}
+```
+
+### `useRef` {/*typing-useref*/}
+
+The [`useRef` hook](/reference/react/useRef) returns a mutable ref object whose `.current` property is initialized to the passed argument. The returned object will persist for the full lifetime of the component.
+
+`useRef` is often used to access the underlying DOM element of a component, but it can also be used to store any mutable value.
+
+When interacting with the DOM, the type of the ref should be set to the type of the underlying DOM element. The naming rule is `HTML` + the name of the element + `Element`, for example `HTMLDivElement` or `HTMLButtonElement`. You can see the full list from TypeScript 5.1 [here](https://github.com/microsoft/TypeScript/blob/a3773ec590c4f0308d546f0e65818cd0d12402f3/src/lib/dom.generated.d.ts#L26899-L27012). These are provided globally by the "DOM" lib, which is included by default in TypeScript projects.
+
+<Sandpack>
+
+```tsx App.tsx active
+import { useRef } from 'react';
+
+export default function Form() {
+  const inputRef = useRef<HTMLButtonElement | null>(null);
+
+  function handleClick() {
+    // The ?. is used because of the `| null` above.
+    inputRef?.current.focus();
+  }
+
+  return (
+    <>
+      <input ref={inputRef} />
+      <button onClick={handleClick}>
+        Focus the input
+      </button>
+    </>
+  );
+}
+```
+
+```js App.js hidden
+import AppTSX from "./App.tsx";
+export default App = AppTSX;
+```
+
+</Sandpack>
+
+
+### `useMemo` / `useCallback` {/*typing-memo-callback*/}
+
+The [`useMemo`](/reference/react/useMemo) and [`useCallback`](/reference/react/useCallback) hooks follow the same pattern with determining their type from the parameter passed to them. If needed you can pass a type argument to them to explicitly set the type.
+
+`useCallback` requires adding your 
+
+### `useEffect` {/*typing-useeffect*/}
+
+The [`useEffect` hook](/reference/react/useEffect) is used to perform side effects in a component. It is called after every render by default, but can be configured to only run when certain values change. The types for this hook allow for either returning a cleanup function or not returning anything.
+
+## Useful Types {/*useful-types*/}
+
+There is quite an expansive set of types which come from the `@types/react` package, it is worth a read when you feel comfortable with how React and TypeScript interact. You can find them [in React's folder in DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts). We will cover a few of the more common types here.
+
+### DOM Events {/*typing-dom-events*/}
+
+When working with DOM events in React, the type of the event is inferred from the event handler. However, when you want to extract a function to be passed to an event handler, you will need to explicitly set the type of the event.
```

Comment:
It would be nice to explain how you choose the `event` type ie. Why `ChangeEvent`? And why `HTMLInputElement`?

Reply:
Yeah, I think this makes sense 👍🏻

Reply:
Extra attention to `target` vs `currentTarget` would be nice since this is a recurring issue where people think `target` should be `T` instead of just `EventTarget` for every event

Reply:
@eps1lon can you please re-explain this point? I'm not quite sure I get the distinction on `target` vs `currentTarget`  enough to get the value

Reply:
We should do this in a follow-up since it depends on the event. It doesn't block this from being merged.

Reply:
Can we juste use the type `const handleChange: ComponentProps<'input'>['onChange'] = (e) => {}` instead manualy typing all args ?